### ISSUE TITLE
Drift detection is flaky

### ISSUE THREAD

**Issue by stouset** [aws/karpenter#3708]
### Version

```
__Karpenter Version:__ v0.25.0
__Kubernetes Version:__ v1.23.17
```

We are not running on the latest version of Karpenter. Reading through the CHANGELOGs, I don't believe versions through 0.27.1 address our issue, but if you have reason to believe otherwise please let me know.

We have enabled drift detection in some of our clusters via the following:

```sh
❯ kubectl --namespace karpenter \
    get configmap karpenter-global-settings \
    --output yaml \
  | yq .data \
  | grep drift

featureGates.driftEnabled: "true"
```

This feature seems extremely brittle. Sometimes it works beautifully, other times it silently does nothing. At times where it's quietly not working, even manually annotating nodes does not cause karpenter to start deprovisioning. Again, nothing relevant is in the logs.

```sh
❯ kubectl annotate node "${node}" 'karpenter.sh/voluntary-disruption=drifted'
```

The only pattern we've noticed is that in working pods, we see the following logs (with some noise removed for clarity):

```
controller.deprovisioning	controller/singleton.go:95	starting controller
controller.deprovisioning	cloudprovider/instancetypes.go:237	discovered EC2 instance types	{"instance-type-count": 626}
controller.deprovisioning	cloudprovider/instancetypes.go:198	discovered EC2 instance types zonal offerings for subnets	{"subnet-selector": "{\"aws-ids\":\"REDACTED,REDACTED,REDACTED\"}"}
...
controller.deprovisioning	deprovisioning/controller.go:145	deprovisioning via drift delete, terminating 1 nodes REDACTED/c6a.8xlarge/on-demand
# nodes continue to deprovision as expected
```

On pods that aren't deprovisioning nodes, we see:

```
controller.deprovisioning	controller/singleton.go:95	starting controller
controller.deprovisioning	cloudprovider/instancetypes.go:237	discovered EC2 instance types	{"instance-type-count": 626}
controller.node-state	cloudprovider/instancetypes.go:198	discovered EC2 instance types zonal offerings for subnets	{"node": REDACTED, "subnet-selector": "{\"aws-ids\":\"REDACTED,REDACTED,REDACTED\"}"}
...
# nothing further
```

Note that in the pod that's detecting drift successfully, discovering EC2 instance types zonal offerings is performed by the deprovisioning controller while in the pod that's stalled, it's being done by the node-state controller. They also produce slightly different log output. I have no idea if this is a clue or a red herring, but it seemed significant enough to warrant mention given the lack of other information I have about this.

Note that both of these pods are configured exactly the same, with the same `configmap/karpenter-global-settings`:

```sh
❯ kubectl --namespace karpenter \
    get configmap karpenter-global-settings \
    --output yaml \
  | yq .data

aws.clusterEndpoint: REDACTED
aws.clusterName: REDACTED
aws.defaultInstanceProfile: REDACTED
aws.enableENILimitedPodDensity: "true"
aws.enablePodENI: "false"
aws.interruptionQueueName: ""
aws.isolatedVPC: "true"
aws.nodeNameConvention: ip-name
aws.vmMemoryOverheadPercent: "0.075"
batchIdleDuration: 1s
batchMaxDuration: 10s
featureGates.driftEnabled: "true"
```

Thanks for investigating!

### Expected Behavior

Drift detection should cause nodes that have drifted to be annotated with `karpenter.sh/voluntary-disruption=drifted`, and notes with that annotation should be deprovisioned.

### Actual Behavior

Nodes that have drifted are not annotated, nor are annotated nodes deprovisioned.

### Steps to Reproduce the Problem

¯\\_(ツ)_/¯

For us, it's intermittent. Sometimes pods perform correctly right off the bat. Sometimes they don't. Sometimes they stop after awhile of working successfully.

### Resource Specs and Logs

Provider:

```yaml
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: normal
spec:
  kubeletConfiguration:
    containerRuntime: dockerd
  labels:
    workload: normal
  limits:
    resources:
      cpu: "9600"
      memory: 19200Gi
  providerRef:
    name: normal
  requirements:
  - key: node.kubernetes.io/instance-type
    operator: In
    values:
    - c6a.8xlarge
    - c6i.8xlarge
    - c5a.8xlarge
    - c5.9xlarge
  - key: kubernetes.io/os
    operator: In
    values:
    - linux
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
  taints:
  - effect: NoSchedule
    key: workload
    value: normal
  ttlSecondsUntilExpired: 604800
```

AWSNodeTemplate:

```yaml
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  name: normal
spec:
  amiFamily: AL2
  blockDeviceMappings:
  - deviceName: /dev/xvda
    ebs:
      deleteOnTermination: true
      encrypted: true
      volumeSize: 500Gi
      volumeType: gp3
  securityGroupSelector:
    aws-ids: REDACTED
  subnetSelector:
    aws-ids: REDACTED
  userData: REDACTED
  securityGroups:
  - id: REDACTED
  - id: REDACTED
  subnets:
  - id: REDACTED
    zone: us-east-1d
  - id: REDACTED
    zone: us-east-1a
  - id: REDACTED
    zone: us-east-1c
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
Are you noticing nodes are annotated with `karpenter.sh/voluntary-disrupted` and aren't deprovisioning or are you just not seeing the nodes get annotated at all?

**Comment by njtran** [comment]
Hey @stouset, so to confirm, you're seeing nodes that are annotated with `karpenter.sh/voluntary-disruption: drifted` not be deprovisioned? Due to how deprovisioning serial nature, you may not see these nodes be deprovisioned immediately, but it should happen eventually if there are no [blocking pods](https://karpenter.sh/v0.27.1/concepts/deprovisioning/#disabling-deprovisioning). If there is a reason a node that you want to be deprovisioned cannot, you should be able to find some `DeprovisioningBlocked` kube-events that describe what may be happening.

Otherwise, it looks like you aren't using an AMISelector, which means that any time a new AL2 EKS Optimized AMI is released, your nodes could be drifted. If a new release happened, you should have seen nodes be annotated and then drifted. Can you confirm this at all?

It looks like the most recent release was on 3/29, which was a week ago. (CLI command referenced below)
```
➜ aws ssm get-parameter --name /aws/service/eks/optimized-ami/1.25/amazon-linux-2/recommended/image_id
{
    "Parameter": {
        "Name": "/aws/service/eks/optimized-ami/1.25/amazon-linux-2/recommended/image_id",
        "Type": "String",
        "Value": "ami-09979d7131171bc02",
        "Version": 4,
        "LastModifiedDate": "2023-03-29T13:58:56.375000-07:00",
        "ARN": "arn:aws:ssm:us-west-2::parameter/aws/service/eks/optimized-ami/1.25/amazon-linux-2/recommended/image_id",
        "DataType": "text"
    }
}
```

**Comment by stouset** [comment]
> so to confirm, you're seeing nodes that are annotated with karpenter.sh/voluntary-disruption: drifted not be deprovisioned?

We are seeing neither drifted nodes get automatically annotated with this, *nor* manually-annotated nodes being deprovisioned.

> Due to how deprovisioning serial nature, you may not see these nodes be deprovisioned immediately, but it should happen eventually if there are no [blocking pods](https://karpenter.sh/v0.27.1/concepts/deprovisioning/#disabling-deprovisioning).

We have waited hours and see no further deprovisions, despite dozens of out-of-spec nodes remaining in the cluster. We also do not have any pods annotated with `karpenter.sh/do-not-evict: "true"`, although we do have *some* pods with [PodDisruptionBudget](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)s that stop them from being moved off nodes. We *do* see karpenter pods with functioning drift detection log when this occurs, after it hard cordoned the node and started moving work off of it. We *do not see* any of this behavior when karpenter's drift detection becomes wedged.

>  If a new release happened, you should have seen nodes be annotated and then drifted. Can you confirm this at all?

During cluster upgrades, we *sometimes* see karpenter pods successfully recognize the new AMI and begin automatic rotation via drift detection. We *more frequently* see no attempts from karpenter to either annotate or deprovision annotated nodes, despite the AMI being out of spec. This issue persists even after a deployment restart. Sometimes the issue suddenly afflicts a running karpenter pod which had previously been successfully rotating nodes.

Karpenter *does* detect the new AMI successfully 100% of time. We have not found an underlying cause that triggers a particular karpenter pod to either fail to perform drift detection on startup, nor have we found a trigger for a functioning karpenter pod to suddenly stop performing drift detection. We have observed both of these behaviors.

This issue occurs independent of if we use `amiFamily` or `amiSelector`. In both scenarios, we often see karpenter pods both failing to annotate drifted nodes and failing to deprovision annotated nodes.

**Comment by njtran** [comment]
> When a karpenter pod enters this "wedged" state (which is a strong majority of the time)

Sorry if I missed this, but what is this "wedged" state? Are you saying the Karpenter pod is crashing or restarting a lot? It sounds like you're saying you run multiple Karpenter pods. How many are you running if you don't mind me asking? I run with two replicas most of the time. 

If you're running Karpenter with many replicas, is there a chance you're watching the logs of the Karpenter pod that isn't leader elected? 

Lastly, just to dot my i's and cross my t's, do you check if a node should be drifted by looking at the AMI label on the node: `karpenter.k8s.aws/instance-ami-id: ami-id`? 

Additionally, you can use kubectl to get the nodes that aren't on the AMI that you want. Can you run this and make sure that it's what you expect the output to be? Something's definitely wrong if the nodes returned from the below command do not have the `karpenter.sh/voluntary-disruption` annotation.
```
    kubectl get node -l karpenter.k8s.aws/instance-ami-id!=${ami-id-you-want}
```

**Comment by stouset** [comment]
> Sorry if I missed this, but what is this "wedged" state? 

This is the entire point of the bug report. Most of the time, karpenter pods sit idly and do perform neither drift detection nor decommissioning of drifted nodes, despite a cluster containing nodes that are out of spec. It emits no logs that would indicate why it is no longer performing these tasks.

Restarting it repeatedly and/or waiting 8-12 hours sometimes results in it resuming its duties.

> It sounds like you're saying you run multiple Karpenter pods. How many are you running if you don't mind me asking? 

Two.

```sh
❯ kubectl --namespace karpenter get pods
NAME                         READY   STATUS    RESTARTS   AGE
karpenter-7c789fc9c4-g9txt   1/1     Running   0          4h32m
karpenter-7c789fc9c4-n4hxl   1/1     Running   0          4h32m
```

**Comment by visokoo** [comment]
I'm seeing this as well in my cluster and the only time I was able to get some drift action was when I delete PDBs (we have some bad ones). Is it possible to have the drift detection controller emit logs for when it can't perform a drift delete because of x reason?

Karpenter Version: 0.27.0
Kubernetes Version: 1.23

**Comment by njtran** [comment]
Thanks @stouset for the comprehensive report! I'm really curious what could get Karpenter stuck here on any of the logic, this definitely shouldn't be happening.

So there are two controllers that are of importance, and I'd like to get down to the problem and see if it's both controllers or just one that's being flaky.
1) Deprovisioning Controller - deprovisions annotated nodes 
2) Node Controller - annotates the node as drifted

For the Deprovisioning Controller, you said:
> The deprovisioner does continue to perform its task of rotating nodes for other reasons such as expiration past their ttlSecondsUntilExpired window.

Which makes me believe that the deprovisioning controller isn't wedged. 

For the Node controller:
It's tough to know how often the node controller is working if you've mixed in annotating some of them yourself. Since it's flaky, it's tough to know by only the information you've given as to when it's worked correctly. 

For both the Deprovisioning and Node controller to not work, I'm thinking there might be some problems with how your settings are persisting, but you said that the issue persists after deployment restarts. We don't load in new settings dynamically, so any time you change the configmap, you'll need to restart your deployment for those configmap changes to take effect. 


The previous release for that region and version was `ami-0f5fb1060cc8f7202` (see output below). Can you de-annotate any of the manually annotated nodes and set your AWSNodeTemplate to the following and wait 10 minutes? 
```
amiFamily: AL2
amiSelector:
    aws-ids: ami-0f5fb1060cc8f7202
```
If you can run the same command to get the AMIs and annotations before changing the AWSNodeTemplate and after 10 minutes, without touching anything, I'd like to see if it works. If not, If you can do `kubectl describe pods karpenter_pod_name -n karpenter` as well. 

```
        {
            "Name": "/aws/service/eks/optimized-ami/1.23/amazon-linux-2/amazon-eks-node-1.23-v20230304",
            "Type": "String",
            "Value": "{\"schema_version\":\"2\",\"image_id\":\"ami-0f5fb1060cc8f7202\",\"image_name\":\"amazon-eks-node-1.23-v20230304\",\"release_version\":\"1.23.16-20230304\"}",
            "Version": 1,
            "LastModifiedDate": "2023-03-09T13:47:35.169000-08:00",
            "ARN": "arn:aws:ssm:us-west-2::parameter/aws/service/eks/optimized-ami/1.23/amazon-linux-2/amazon-eks-node-1.23-v20230304",
            "DataType": "text"
        },
```


---

Sorry to ask so much, but it's tough to debug flakes since they're especially hard to reproduce.

**Comment by stouset** [comment]
> Which makes me believe that the deprovisioning controller isn't wedged.

What makes me dispute this is the fact that manually-annotated nodes aren't being deprovisioned, but we'll run this test and see!

> The previous release for that region and version was `ami-0f5fb1060cc8f7202` (see output below). Can you de-annotate any of the manually annotated nodes and set your AWSNodeTemplate to the following and wait 10 minutes?

```sh
❯ kubectl get nodes --no-headers | awk '{print $1}' | while read node
    kubectl annotate node $node 'karpenter.sh/voluntary-disruption-'
end

❯ kubectl get nodes \
    --no-headers \
    --output jsonpath='{range .items[*]}{.metadata.labels.karpenter\.k8s\.aws/instance-ami-id}{" "}{.metadata.annotations.karpenter\.sh/voluntary-disruption}{"\n"}{end}' \
  | sort \
  | uniq -c
  26 ami-0a88006b97dc624e7 
```

I'm experimenting in us-west-2, so I'll use AMI `ami-0e968e66a05cef9f9`, which is one previous from the current one in this region.

```sh
❯ kubectl --namespace karpenter get awsnodetemplates --output json | jq '.spec | {amiFamily, amiSelector}'
{
  "amiFamily": "AL2",
  "amiSelector": {
    "aws-ids": "ami-0e968e66a05cef9f9"
  }
}

❯ kubectl --namespace karpenter rollout restart deployment karpenter
deployment.apps/karpenter restarted
```

Immediately afterward, most of the nodes are detected as drifted, and one new node has been launched.

```sh
❯ kubectl get nodes \
    --no-headers \
    --output jsonpath='{range .items[*]}{.metadata.labels.karpenter\.k8s\.aws/instance-ami-id}{" "}{.metadata.annotations.karpenter\.sh/voluntary-disruption}{"\n"}{end}' \
  | sort \
  | uniq -c
   4 ami-0a88006b97dc624e7 
  22 ami-0a88006b97dc624e7 drifted
   1 ami-0e968e66a05cef9f9 
```

This is consistent with what I've seen. We'll see whether or not the remaining `ami-0a88006b97dc624e7` nodes are annotated with drift, or whether all of the `ami-0a88006b97dc624e7` are rotated. I predict no :)

> If you can run the same command to get the AMIs and annotations before changing the AWSNodeTemplate and after 10 minutes, without touching anything, I'd like to see if it works. If not, If you can do `kubectl describe pods karpenter_pod_name -n karpenter` as well.

What would you like to see from this output?

> Sorry to ask so much, but it's tough to debug flakes since they're especially hard to reproduce.

Ask everything you need in order to help identify this issue!

**Comment by tzneal** [comment]
We discovered an issue where Karpenter would start to backoff upon successful deprovisioning.  I believe that this is what you are seeing due to:

```
 kubectl --namespace karpenter rollout restart deployment karpenter
deployment.apps/karpenter restarted

Immediately afterward, most of the nodes are detected as drifted, and one new node has been launched.
```

This restart of Karpenter resets the backoff, where it starts to replace the drifted nodes and backoff starts to accumulate again.

**Comment by stouset** [comment]
Can you explain in a little further detail?

**Comment by tzneal** [comment]
For the bug that we're currently aware of, restarting should always resolve it. However it will immediately reoccur (e.g. replace a few nodes with more time between each replacement appearing to eventually stop).

**Comment by stouset** [comment]
This is potentially part of the problem, as I have noticed karpenters rotate until grinding to a halt. But there may be another issue as well, as restarting does not always prompt karpenter to continue.

**Comment by ellistarn** [comment]
This label is deprecated, btw: https://github.com/aws/karpenter/blob/main/pkg/apis/v1alpha1/register.go#L83. Looking at the code, it has no bearing on drift.

**Comment by ellistarn** [comment]
We suspect that this is root caused by https://github.com/aws/karpenter-core/pull/279/files. Though, the fact that restart doesn't solve it in all cases is concerning. Will be coming out next release..

**Comment by stouset** [comment]
Great. We'll update once this is released and can pursue further if there continue to be issues.

**Comment by tzneal** [comment]
This should be resolved in v0.27.2, please re-open or file a new issue if you still experience this.

**Comment by stouset** [comment]
This issue still seems to be happening. However, I now see this in the logs:

```
2023-04-14T19:19:37.349Z	DEBUG	controller.deprovisioning	waiting on cluster sync	{"commit": "d01ea11-dirty"}
```

No further `controller.deprovisioning` messages occur. In this cluster, the out-of-spec nodes have been annotated as `drifted` already.

```sh
❯ kubectl get nodes \
    -ojsonpath='{range .items[*]}{.metadata.labels.workload}{" "}{.metadata.labels.ami_id}{.metadata.labels.karpenter\.k8s\.aws/instance-ami-id}{" "}{.status.nodeInfo.kubeletVersion}{" "}{.metadata.annotations.karpenter\.sh/voluntary-disruption}{"\n"}{end}' \
  | sort \
  | uniq -c

   3 gpu-nvidia-v1 ami-02c5235360fb8957c v1.23.17-eks-a59e1f0 
  16 normal ami-04c545ca749cfe57b v1.23.17-eks-a59e1f0 drifted
 122 normal ami-0559f43d314aa2e7e v1.23.17-eks-a59e1f0 
   5 pci ami-0559f43d314aa2e7e v1.23.17-eks-a59e1f0 
   9 utility ami-0559f43d314aa2e7e v1.23.17-eks-a59e1f0 
```

**Comment by tzneal** [comment]
If you describe the node, are there events on it?

**Comment by tzneal** [comment]
Closing again due to #3764
